src/java.net.http/share/classes/jdk/internal/net/http/Http2Connection.java

```diff
@@ -864,9 +868,19 @@ private <T> void handlePushPromise(Stream<T> parent, PushPromiseFrame pp)
         // always decode the headers as they may affect connection-level HPACK
         // decoding state
         assert pushContinuationState == null;
+        int promisedStreamid = pp.getPromisedStream();
+        if (promisedStreamid != nextPushStream) {
+            resetStream(promisedStreamid, ResetFrame.PROTOCOL_ERROR);
+            return;
+        } else if (!reserveStream(false)) {
+            resetStream(promisedStreamid, ResetFrame.REFUSED_STREAM);
+            return;
+        } else {
+            nextPushStream += 2;
+        }
+
```

Comment:
Hi Conor - sorry for suggesting to move this code up into `handlePushPromise`. Now that I see the whole picture I believe these lines should have stayed where they were, in `completePushPromise`. The main reason is that we need to properly identify and decode possible ContinuationFrames as being part of the push promise in order to maintain the HPack encoders/decoders in sync, and we can only do that if we keep the `pushContinuationState` in place until we have received the END_HEADERS flag.

Sorry for suggesting that wrong move - it was my mistake.

Reply:
No problem at all Daniel. I see what you're saying, essentially that if the Continuation does come in on the wrong stream id, it still needs to be decoded and identified as a part of the Push Promise before being rejected with the wrong stream id.